url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/core/CellIntegration.java

```diff
@@ -0,0 +1,270 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package de.jannlab.core;
+
+/**
+ * This class contains static methods for computing serveral 
+ * integrations function. An integration function "integrates" 
+ * data from a source array into a destination array in a 
+ * specific way. Which value are integrated and where the results
+ * are stored is determined by a sequence of links. Thereby, each
+ * source value is multiplied with the weight of the link (an
+ * non-weighted link has weight 1.0). The overhead of multiplying
+ * 1.0 on non-weighted links is much faster than an "if"-check for
+ * each weight.
+ * <br></br>
+ * @see Link
+ * @author Sebastian Otte
+ *
+ */
+public final class CellIntegration {
+    //
+    /** None integration function. */
+    public static final int NONE    = 0;
+    /** Ordinary sum over all values. */
+    public static final int SUM     = 1;
+    /** Product over all values. */
+    public static final int MULT    = 2;
+    /** Copy last source value for dest. */
+    public static final int LASTID = 3;
+    //
+    
+    /**
+     * This method computes the weighted sum for the given arguments.
+     * <br></br>
+     * @param src Refers the source data array.
+     * @param dst Refers the destination data array.
+     * @param cellsoff Refers the lower bound of the destination range.
+     * @param cellsnum Refers the size of the destination range.
+     * @param weights Refers the weights vector.
+     * @param links Refers the links vector.
+     * @param linksoff Gives the offset of the first link according 
+     * the this current context.
+     * @param linksnum Gives the number of links according to the current context. 
+     */
+    public static void sum(
+            final double[] src,
+            final double[] dst,
+            final int cellsoff,
+            final int cellsnum,
+            final double[] weights,
+            final int[] links,
+            final int linksoff,
+            final int linksnum
+    ) {
+        if (linksnum == 0) return;
+        //
+        int link = linksoff;
+        //
+        // clean cell inputs.
+        //
+        int end = (cellsoff + cellsnum);
+        for (int i = cellsoff; i < end; i++) {
+            dst[i] = 0.0;
+        }   
+        //
+        // compute weighted sum over all links.
+        //
+        for (int i = 0; i < linksnum; i++) {
+            //
+            final int ci = links[link + Link.IDX_SRC]; 
+            final int cj = links[link + Link.IDX_DST];
+            final int ij = links[link + Link.IDX_WEIGHT];
+            ///
+            final double xi  = src[ci];
+            final double wij = weights[ij];
+            //
+            dst[cj] += (xi * wij);
+            //
+            // next link.
+            //
+            link += Link.LINK_SIZE;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 71:98, 170:205.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.